docs/5 - Parent and child components.md

```diff
@@ -1,3 +1,473 @@
 # Parent and child components
 
-Chapter coming very soon!
+So far the examples have only concerned a single component, however this will only take us so far before the state and query algebra becomes unmanageable. The answer is to break our app into components that can be composed.
+
+Let's take a look at a component that uses our button component as a child:
+
+``` purescript
+import Prelude
+import Data.Maybe (Maybe(..), maybe)
+import Halogen as H
+import Halogen.HTML as HH
+import Halogen.HTML.Events as HE
+import Button as Button
+
+data Query a
+  = HandleButton Button.Message a
+  | CheckButtonState a
+
+type State =
+  { toggleCount :: Int
+  , buttonState :: Maybe Boolean
+  }
+
+data Slot = ButtonSlot
+derive instance eqButtonSlot :: Eq Slot
+derive instance ordButtonSlot :: Ord Slot
+
+component :: forall m. H.Component HH.HTML Query Unit Void m
+component =
+  H.parentComponent
+    { initialState: const initialState
+    , render
+    , eval
+    , receiver: const Nothing
+    }
+  where
+
+  initialState :: State
+  initialState =
+    { toggleCount: 0
+    , buttonState: Nothing }
+
+  render :: State -> H.ParentHTML Query Button.Query Slot m
+  render state =
+    HH.div_
+      [ HH.slot ButtonSlot Button.myButton unit (HE.input HandleButton)
+      , HH.p_
+          [ HH.text ("Button has been toggled " <> show state.toggleCount <> " time(s)") ]
+      , HH.p_
+          [ HH.text
+              $ "Last time I checked, the button was: "
+              <> (maybe "(not checked yet)" (if _ then "on" else "off") state.buttonState)
+              <> ". "
+          , HH.button
+              [ HE.onClick (HE.input_ CheckButtonState) ]
+              [ HH.text "Check now" ]
+          ]
+      ]
+
+  eval :: Query ~> H.ParentDSL State Query Button.Query Slot Void m
+  eval = case _ of
+    HandleButton (Button.Toggled _) next -> do
+      H.modify (\st -> st { toggleCount = st.toggleCount + 1 })
+      pure next
+    CheckButtonState next -> do
+      buttonState <- H.query ButtonSlot $ H.request Button.IsOn
+      H.modify (_ { buttonState = buttonState })
+      pure next
+```
+
+A runnable version of this is available in the [`components` example](../examples/components/).

```

Comment:
I think it's better to start with the example, but it should probably mention that the main difference is that it's using `parentComponent`, but otherwise the spec as the same arguments.